code/modules/antagonists/eldritch_cult/eldritch_effects.dm

```diff
@@ -0,0 +1,213 @@
+		qdel(X)
+	for(var/Y in targets)
+		targets -= Y
+	return ..()
+
+/**
+  * Automatically fixes the target and smash network
+  *
+  * Fixes any bugs that are caused by late Generate() or exchanging clients
+  */
+/datum/reality_smash_tracker/proc/ReworkNetwork()
+	for(var/mind in targets)
+		for(var/X in smashes)
+			var/obj/effect/reality_smash/reality_smash = X
+			reality_smash.AddMind(mind)
+
+/**
+  * Generates a set amount of reality smashes based on the N value
+  *
+  * Automatically creates more reality smashes
+  */
+/datum/reality_smash_tracker/proc/Generate(var/n)
+	var/number = round(max(n * 6,10),1)
+
+	for(var/i in 0 to number)
+
+		var/obj/effect/landmark/L = pick(GLOB.generic_event_spawns + GLOB.landmarks_list)
+		var/obj/effect/reality_smash/RS = new(L.drop_location())
+		smashes += RS
+
+	ReworkNetwork()
+
+/**
+  * Adds a mind to the list of people that can see the reality smashes
+  *
+  * Use this whenever you want to add someone to the list
+  */
+/datum/reality_smash_tracker/proc/AddMind(var/datum/mind/M)
+	RegisterSignal(M.current,COMSIG_MOB_LOGIN,.proc/ReworkNetwork)
+	targets |= M
+	for(var/X in smashes)
+		var/obj/effect/reality_smash/reality_smash = X
+		reality_smash.AddMind(M)
+
+/**
+  * Removes a mind from the list of people that can see the reality smashes
+  *
+  * Use this whenever you want to remove someone from the list
+  */
+/datum/reality_smash_tracker/proc/RemoveMind(var/datum/mind/M)
+	targets -= M
+	for(var/obj/effect/reality_smash/RS in smashes)
+		RS.RemoveMind(M)
+
+/obj/effect/broken_illusion
+	name = "Pierced reality"
+	icon = 'icons/effects/eldritch.dmi'
+	icon_state = "pierced_illusion"
+
+/obj/effect/broken_illusion/examine(mob/user)
+	if(!IS_HERETIC(user) && ishuman(user))
+		var/mob/living/carbon/human/human_user = user
+		to_chat(human_user,"<span class='warning'>Your brain hurts when you look at this!</span>")
+		human_user.adjustOrganLoss(ORGAN_SLOT_BRAIN,10)
+	. = ..()
+
+/obj/effect/reality_smash
+	name = "/improper reality smash"
+	icon = 'icons/effects/eldritch.dmi'
+	///We cannot use icon_state since this is invisible, functions the same way but with custom behaviour.
+	var/image_state = "reality_smash"
+	///Who can see us?
+	var/list/minds = list()
+	///Tracked image
+	var/image/img
+
+/obj/effect/reality_smash/Initialize()
+	. = ..()
+	img = image(icon, src, image_state, OBJ_LAYER)
+	generate_name()
+
+/obj/effect/reality_smash/Destroy()
+
+	for(var/cm in minds)
+		var/datum/mind/cultie = cm
+		if(cultie.current?.client)
+			cultie.current.client.images -= img
+		//clear the list
+		minds -= cultie
+	img = null
+	new /obj/effect/broken_illusion(drop_location())
+	return ..()
+
+///Makes the mind able to see this effect
+/obj/effect/reality_smash/proc/AddMind(var/datum/mind/cultie)
+	if(cultie.current.client)
+		minds |= cultie
+		cultie.current.client.images |= img
+
+///Makes the mind not able to see this effect
+/obj/effect/reality_smash/proc/RemoveMind(var/datum/mind/cultie)
```

Comment:
This whole tracking is asking for edge cases, i think it would be more sensible to let antag datum subscribe to login and handle showing/hiding these than the objects themselves.

Reply:
i wanted to contain all code that handles generation , showing and hiding in a single datum, having it split between antag datum would make the code less readable,also there is a special proc that fixes the network if something goes wrong.